packages/ubuntu_desktop_installer/lib/pages/who_are_you/who_are_you_page.dart

```diff
@@ -0,0 +1,368 @@
+import 'package:encrypt/encrypt.dart' as encrypt;
+import 'package:flutter/material.dart';
+import 'package:form_field_validator/form_field_validator.dart';
+import 'package:provider/provider.dart';
+import 'package:subiquity_client/subiquity_client.dart';
+import 'package:yaru/yaru.dart' as yaru;
+
+import '../../routes.dart';
+import '../../widgets/localized_view.dart';
+import '../../widgets/validated_input.dart';
+import '../wizard_page.dart';
+import 'who_are_you_model.dart';
+
+/// The installer page for setting up the user data.
+///
+/// It uses [WizardPage] and [WizardAction] to create an installer page.
+class WhoAreYouPage extends StatefulWidget {
+  /// Creates a the installer page for setting up the user data.
+  const WhoAreYouPage({Key? key}) : super(key: key);
+
+  ///
+  static Widget create(BuildContext context) {
+    return ChangeNotifierProvider(
+      create: (_) => WhoAreYouModel(),
+      child: WhoAreYouPage(),
+    );
+  }
+
+  @override
+  _WhoAreYouPageState createState() => _WhoAreYouPageState();
+}
+
+class _WhoAreYouPageState extends State<WhoAreYouPage> {
+  late WhoAreYouModel _whoAreYouModel;
+  final _realNameController = TextEditingController();
+  final _computerNameController = TextEditingController();
+  final _usernameController = TextEditingController();
+  final _passwordController = TextEditingController();
+  final _confirmPasswordController = TextEditingController();
+
+  final _usernameFormKey = GlobalKey<FormState>();
+  final _realNameFormKey = GlobalKey<FormState>();
+  final _computerNameFormKey = GlobalKey<FormState>();
+  final _passwordFormKey = GlobalKey<FormState>();
+  final _confirmPasswordFormKey = GlobalKey<FormState>();
+
+  @override
+  void initState() {
+    super.initState();
+
+    _whoAreYouModel = Provider.of<WhoAreYouModel>(context, listen: false);
+    _whoAreYouModel.loadProfileSetup().then((_) {
+      _usernameController.text = _whoAreYouModel.username;
+    });
+
+    _realNameController.addListener(() {
+      _whoAreYouModel.realName = _realNameController.text;
+    });
+
+    _computerNameController.addListener(() {
+      _whoAreYouModel.hostName = _computerNameController.text;
+    });
+
+    _usernameController.addListener(() {
+      _whoAreYouModel.username = _usernameController.text;
+    });
+    _passwordController.addListener(() {
+      _whoAreYouModel.password = _passwordController.text;
+    });
+  }
+
+  @override
+  void dispose() {
+    _realNameController.dispose();
+    _computerNameController.dispose();
+    _usernameController.dispose();
+    _passwordController.dispose();
+    _confirmPasswordController.dispose();
+    super.dispose();
+  }
+
+  Widget findPassWordStrengthLabel(
```

Comment:
- add `WhoAreYouModel.passwordStrength`
- determine the strength in the model instead as it's "business logic"
- return an enum value from the model, indicating the strength
- here in the view, build the representation (widget) for the strength returned from the model

Reply:
i don't think this is a good idea, since the validation is only connected to the label "tip"
the current ubuntu installer validation only requires 2 digits, then you are allowed to proceed.

Reply:
i'm not sure i follow. :smile: the basic principle is to have functionality in the model and keep the view as a declarative presentation. you can keep the label creation here, just move the logic into the model so we can test it in a sensible way (not as a slow widget test, but as a unit test that spams the model with a combination of different password inputs and validates that the result is what we expect it to be). also, changing the rules later should not affect the UI as it's just a presentation of weak/good/strong password but it doesn't care how it's determined

Reply:
even if the validation requires only 2 characters, isn't the password strength going to be visualized as weak/good/strong? this is orthogonal to the validation logic, right?

P.S. I would very much like to have the actual validation logic out of the view too, for the very same purposes as described above, but let's take it step by step.

Reply:
there are basically too validations going on here:

1. a bare minimum validation needed to proceed with the page, like "required" or "2 digits", which is in sync with the current ubuntu installer, so if this should be stronger we need someone from... design? so say something about this
2. a second validation _only_ needed for this "I am your password assistant telling you how good your password is but won't stop you to proceed with a bad password"-hint

Why should the second hinting be used inside the model? It is not relevant for the model, unless the design? wants us to force the user to make a better password to proceed

Reply:
soooo the first validation could be indeed used to the model, but the second not, in my humble opinion ;)

Reply:
what i'm trying to say is that the responsibility of this function should be _visualizing_ "weak" "good" or "strong", not visualizing _and_ determining the strength.

> there are basically 2 validations going on here:

the validation rules are irrelevant for determining the strength of a password, right?  in other words, the strength does not affect the validation result. this is a standalone function, not strictly connected to the validation which is done elsewhere outside this function.

> Why should the second validation for hinting be used inside the model? It is not relevant for the model, unless the design?

if you mean software design, yes the view model is there specifically to help keep the view as simple, clean, and declarative as possible.

> wants us to force the user to make a better password to proceed

if the strength would affect the continue-button, that should be determined by the view model too. :wink:

Reply:
right, I guess this is rather a question of taste... I would personally detach the model from the view completely. I will do as you suggested

Reply:
i have a feeling that the term "model" is the source of confusion here.

- in MVVM, which we are using, view model is like a relaxed version of MVC's controller
- some use the model term for data classes too (view model vs. data model)
- others prefer to call them just data classes (like the data classes in subiquity_client)
- some UI frameworks use the term "model" for item view's data source (lists, grids, tables...)
- probably many others...

is it possible that we are mixing up view models (business logic) with data models (pure data blob)?

Reply:
also there can not be a global "model.isValid ?" like in the WSL page you created because each field could potentially have different validations. 
ontop there are multiple messages for those each field, so I would need to add a function for each of the wrong values like
you can not simple say "whoAreYouModel.isPasswordValid" because this axes again other options and thus other error texts

```dart
validator: (_) => whoAreYouModel.isPasswordEmpty
                                        ? lang.whoAreYouPagePasswordRequiredValidatorErrorText
                                        : whoAreYouModel.hasPasswordTheMinLengh ? lang.whoAreYouPagePasswordRequiredValidatorErrorText : null
```


```dart
validator: MultiValidator([
    RequiredValidator(
        errorText: lang
            .whoAreYouPagePasswordRequiredValidatorErrorText),
    MinLengthValidator(2,
        errorText: lang
            .whoAreYouPagePasswordMinLengthValidatorErrorText),
  ])
```

Do you really think this is an improvement? :thinking: @jpnurmi

Reply:
and the error text string values must come from the lang object, which the model does not know

Reply:
yes, the model should not care about UI strings or widgets. the model knows the password, so it's a matter of querying the strength from the model (describe it as enum)

Reply:
Pseudo example:
```
enum PasswordStrength { weak, good, strong }

class WhoAreYouModel extends ChangeNotifier {
  // ...
  
  PasswordStrength? get passwordStrength {
    const strongPassword = RegExp(...);
    const goodPassword = RegExp(...);

    if (strongPassword.hasMatch(password)) {
      return PasswordStrength.strong;
    }
    if (goodPassword.hasMatch(password)) {
      return PasswordStrength.good;
    }
    if (password.length < something) {
      return PasswordStrength.weak;
    }
    return null;
  }
}
```

Reply:
the respective view would be something like:
```dart
switch (model.passwordStrength) {
  case PasswordStrength.weak: return Text(lang.foo, style: bar);
  case ...
  default: return SizedBox.shrink();
}
```

Reply:
notice that the only difference is that the logic and the presentation are separate. two functions, each with a clear responsibility, instead of one function doing both. :smile: 

not everyone is fluent in regexp patterns, so it's a good practice to test them to catch potential regressions when the rules are tweaked later. this is straightforward when all you need is to create a model instance, feed it with a ton of different passwords, and test that the result is as expected. this works great as a plain unit test. if it was done as a widget test, the test would have to create an instance of the whole page UI, find the correct input field instance, ensure that it has input focus, emulate entering text, find the result label, and then introspect its text.